### PR TITLE
support for package debugging

### PR DESCRIPTION
**Pull request by rohit21agrawal**
Fixes: https://github.com/NuGet/Home/issues/6878

Adds support in Pack for creating the new symbols package (snupkg) and support in push to be able to push the same by discovering whether the endpoint supports the capability.

Spec for the same can be found at: https://github.com/NuGet/Home/wiki/NuGet-Client-Package-Debugging
and https://github.com/NuGet/Home/wiki/NuGet-Package-Debugging-&-Symbols-Improvements

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs [hunk 1]
```diff
@@ -106,14 +109,20 @@ public override void ExecuteCommand()
...
                 if (!System.Version.TryParse(MinClientVersion, out _minClientVersionValue))
                 {
                     throw new CommandLineException(LocalizedResourceManager.GetString("PackageCommandInvalidMinClientVersion"));
                 }
             }
 
+            if(!string.IsNullOrEmpty(SymbolPackageFormat))
+            {
+                packArgs.SymbolPackageFormat = PackArgs.GetSymbolPackageFormat(SymbolPackageFormat);
+            }
```
Comment (loic-sharma) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs:123]: >} [](start = 12, length = 1)

You have two empty lines below this one

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 2]
```diff
@@ -1240,6 +1258,13 @@ private void AddFiles(Packaging.PackageBuilder builder, string itemType, string
                     continue;
                 }
 
+                if (IncludeSymbols &&
+                    SymbolPackageFormat == SymbolPackageFormat.Snupkg &&
+                    !string.Equals(Path.GetExtension(fullPath), ".pdb", StringComparison.OrdinalIgnoreCase))
```
Comment (loic-sharma) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:1263]: >StringComparison.OrdinalIgnoreCase [](start = 72, length = 34)

Should this be using `PathUtility.GetStringComparisonBasedOnOS`?

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 3]
```diff
@@ -979,9 +978,17 @@ internal static void ExcludeFilesForLibPackage(ICollection<IPackageFile> files)
             PathResolver.FilterPackageFiles(files, file => file.Path, _libPackageExcludes);
         }
 
-        internal static void ExcludeFilesForSymbolPackage(ICollection<IPackageFile> files)
+        internal static void ExcludeFilesForSymbolPackage(ICollection<IPackageFile> files, SymbolPackageFormat symbolPackageFormat)
         {
             PathResolver.FilterPackageFiles(files, file => file.Path, _symbolPackageExcludes);
+            if(symbolPackageFormat == SymbolPackageFormat.Snupkg)
+            {
+                var toRemove = files.Where(t => !string.Equals(Path.GetExtension(t.Path), ".pdb", StringComparison.OrdinalIgnoreCase)).ToList();
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:986]: >OrdinalIgnoreCase [](start = 115, length = 17)

Should this use `PathUtility.GetStringComparisonBasedOnOS`?

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs [hunk 4]
```diff
@@ -149,11 +152,23 @@ private void AddOutputLibFiles(IEnumerable<OutputLibFile> libFiles, HashSet<stri
...
         {
             if (!Files.Any(p => packageFile.Target.Equals(p.Target, StringComparison.CurrentCultureIgnoreCase)))
             {
-                Files.Add(packageFile);
+                var fileExtension = Path.GetExtension(packageFile.Source);
+
+                if(IncludeSymbols &&
+                    PackArgs.SymbolPackageFormat == SymbolPackageFormat.Snupkg &&
+                    !string.Equals(fileExtension, ".pdb", PathUtility.GetStringComparisonBasedOnOS()))
+                {
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: It may be worth adding a comment as to why the `StringComparison` matters here depending on the OS.

File: src/NuGet.Core/NuGet.Protocol/Resources/SymbolPackageUpdateResourceV3.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using NuGet.Common;
+
+namespace NuGet.Protocol.Core.Types
+{
+    /// <summary>
+    /// Contains logics to push or delete packages in Http server or file system
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Resources/SymbolPackageUpdateResourceV3.cs]: >packages [](start = 42, length = 8)

`packages` -> `symbols packages`?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 2 above)
Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:1263]: this is an extension comparison, and extension in any casing would imply the same.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 3 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:986]: same here. pdb extension in any casing would imply symbols.

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs (hunk 4 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: it doesn't actually. i'll change to ordinalignorecase.

**Comment by rohit21agrawal** [comment]
@NuGet/nuget-client 🔔

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets [hunk 6]
```diff
@@ -39,12 +39,14 @@ Copyright (c) .NET Foundation. All rights reserved.
     <PackDependsOn>$(BeforePack); _IntermediatePack; GenerateNuspec; $(PackDependsOn)</PackDependsOn>
     <IsInnerBuild Condition="'$(TargetFramework)' != '' AND '$(TargetFrameworks)' != ''">true</IsInnerBuild>
     <NoBuild Condition="'$(GeneratePackageOnBuild)' == 'true'">true</NoBuild>
+    <SymbolPackageFormat Condition="'$(SymbolPackageFormat)' == ''">symbols.nupkg</SymbolPackageFormat>
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:42]: Isn't the default `snupkgs`?

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets [hunk 7]
```diff
@@ -39,12 +39,14 @@ Copyright (c) .NET Foundation. All rights reserved.
...
     <ImportNuGetBuildTasksPackTargetsFromSdk Condition="'$(ImportNuGetBuildTasksPackTargetsFromSdk)' == ''">false</ImportNuGetBuildTasksPackTargetsFromSdk>
     <AllowedOutputExtensionsInPackageBuildOutputFolder>.dll; .exe; .winmd; .json; .pri; .xml; $(AllowedOutputExtensionsInPackageBuildOutputFolder)</AllowedOutputExtensionsInPackageBuildOutputFolder>
-    <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>.pdb; .mdb; $(AllowedOutputExtensionsInPackageBuildOutputFolder); $(AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder)</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>
+    <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder Condition="'$(SymbolPackageFormat)' != 'snupkg'">.pdb; .mdb; $(AllowedOutputExtensionsInPackageBuildOutputFolder); $(AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder)</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>
+    <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder Condition="'$(SymbolPackageFormat)' == 'snupkg'">.pdb</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:49]: since a `snupkgs` will also have its own `nuspec` what happens customer wants to add additional files in there like they have the flexibility to extend in `symbol.nupkg`...

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 8]
```diff
@@ -804,9 +813,18 @@ private void AddOutputFiles(Packaging.PackageBuilder builder)
...
             if (IncludeSymbols)
             {
-                outputFileNames.Add($"{targetFileName}.pdb");
-                outputFileNames.Add($"{targetFileName}.dll.mdb");
-                outputFileNames.Add($"{targetFileName}.exe.mdb");
+                // if this is a snupkg package, we don't want any files other than symbol files.
+                if(SymbolPackageFormat == SymbolPackageFormat.Snupkg)
+                {
+                    outputFileNames.Clear();
+                    outputFileNames.Add($"{targetFileName}.pdb");
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:820]: how are you enabling source linking? or is that not part of this PR? if not, then probably mention what all this PR covers out of original spec.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 3 above)
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:986]: We should log some warning when we're filtering these files

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 9]
```diff
@@ -1014,22 +1021,24 @@ public static string GetOutputPath(PackageBuilder builder, PackArgs packArgs, bo
                 }
             }
 
-            var outputFile = GetOutputFileName(builder.Id, versionToUse, isNupkg: isNupkg, symbols: symbols, excludeVersion: packArgs.OutputFileNamesWithoutVersion);
+            var outputFile = GetOutputFileName(builder.Id, versionToUse, isNupkg: isNupkg,
+                symbols: symbols, symbolPackageFormat: packArgs.SymbolPackageFormat,
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: style: have each parameter in separate line

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs [hunk 10]
```diff
@@ -36,6 +37,21 @@ public static class PushRunner
             }
 
             var packageUpdateResource = await CommandRunnerUtility.GetPackageUpdateResource(sourceProvider, source);
+            SymbolPackageUpdateResourceV3 symbolPackageUpdateResource = null;
+            // figure out from index.json if pushing snupkg is supported
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs]: nit: blank line before comment

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs [hunk 11]
```diff
@@ -36,6 +37,21 @@ public static class PushRunner
...
+            if (string.IsNullOrEmpty(symbolSource)
+                && !noSymbols
+                && !sourceUri.IsFile
+                && sourceUri.IsAbsoluteUri)
+            {
+                symbolPackageUpdateResource = await CommandRunnerUtility.GetSymbolPackageUpdateResource(sourceProvider, source);
+                if (symbolPackageUpdateResource != null)
+                {
+                    symbolSource = symbolPackageUpdateResource.SourceUri.AbsoluteUri;
+                    symbolApiKey = apiKey;
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs:53]: why a separate `symbolApiKey`?

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs [hunk 12]
```diff
@@ -68,20 +69,48 @@ public Uri SourceUri
...
+        {
+            await Push(
+                packagePath,
+                symbolSource,
+                timeoutInSecond,
+                disableBuffering,
+                getApiKey,
+                getSymbolApiKey,
+                noServiceEndpoint,
+                null,
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs]: style: named parameter

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs [hunk 13]
```diff
@@ -122,15 +151,17 @@ public Uri SourceUri
...
             CancellationToken token)
         {
+            var isSymbolEndpointSnupkgCapable = symbolPackageUpdateResource != null;
             // Get the symbol package for this package
-            var symbolPackagePath = GetSymbolsPath(packagePath);
+            var symbolPackagePath = GetSymbolsPath(packagePath, isSymbolEndpointSnupkgCapable);
 
             // Push the symbols package if it exists
-            if (File.Exists(symbolPackagePath))
+            if (File.Exists(symbolPackagePath) || symbolPackagePath.IndexOf('*') != -1)
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs:164]: what does `*` means here?

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs [hunk 14]
```diff
@@ -122,15 +151,17 @@ public Uri SourceUri
             string source,
             string apiKey,
             bool noServiceEndpoint,
+            SymbolPackageUpdateResourceV3 symbolPackageUpdateResource,
             TimeSpan requestTimeout,
             ILogger log,
             CancellationToken token)
         {
+            var isSymbolEndpointSnupkgCapable = symbolPackageUpdateResource != null;
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs:159]: are we only pushing `snupkgs` now? or do we still allow pushing `symbol.nupkg`?

**Review by jainaashish — COMMENTED** [review]
There were couple of things which were not clear from spec or implementation:
- will we still be publishing symbol.nupkg?
- what happen when there is a symbol.nupkg but server supports snupkg?
- will we still be publishing symbol.nupkg to SymbelServer by default?

**Comment by rohit21agrawal** [comment]
@jainaashish to answer your questions:

> will we still be publishing symbol.nupkg?

symbol.nupkg are not published by default even now, until a symbol source and api key has been specified explicitly. we will keep that behavior intact.

> what happen when there is a symbol.nupkg but server supports snupkg?

like i said above, if there is a symbol.nupkg, the only way it will be pushed is if a symbol source and apikey is specified explicitly. If that is specified explicitly, we will look for symbol.nupkg instead of snupkg. For pushing a snupkg, you do not need to specify symbol source or an api key.

> will we still be publishing symbol.nupkg to SymbelServer by default?

we don't do this today, and we aren't going to change that.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs (hunk 13 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs:164]: you can specify *.nupkg in push command.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 3 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:986]: no, this is by design, so no need to log warnings.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 8 above)
Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:820]: source linking is not something we enable . we only care about packing the PDBs. it's the authors responsibility to sourcelink or not. regardless, it's not a requirement.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets (hunk 7 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:49]: snupkg should not have any files other than PDBs . that is by design.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets (hunk 6 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:42]: no. we will change the default in 5.0 at some time.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:42]: The spec needs to be aligned to the behavior then. 
The spec says otherwise https://github.com/NuGet/Engineering/blob/master/Client.Specs/PackageDebugging.md#package-authoring-experience

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 15]
```diff
@@ -302,7 +305,11 @@ public Packaging.PackageBuilder CreateBuilder(string basePath, NuGetVersion vers
             // Add sources if this is a symbol package
             if (IncludeSymbols)
             {
-                ApplyAction(p => p.AddFiles(builder, SourcesItemType, SourcesFolder));
+                if (SymbolPackageFormat != SymbolPackageFormat.Snupkg)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs]: Why not do a positive check? 
A potential 3rd option would break this in the future. 
`SymbolPackageFormat == SymbolPackageFormat.SymbolsNupkg`

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 16]
```diff
@@ -804,9 +813,18 @@ private void AddOutputFiles(Packaging.PackageBuilder builder)
...
-                outputFileNames.Add($"{targetFileName}.pdb");
-                outputFileNames.Add($"{targetFileName}.dll.mdb");
-                outputFileNames.Add($"{targetFileName}.exe.mdb");
+                // if this is a snupkg package, we don't want any files other than symbol files.
+                if(SymbolPackageFormat == SymbolPackageFormat.Snupkg)
+                {
+                    outputFileNames.Clear();
+                    outputFileNames.Add($"{targetFileName}.pdb");
+                }
+                else
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:822]: If you're gonna have an enum, you can use a switch to make it a tiny bit more future proof.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx [hunk 17]
```diff
@@ -5480,4 +5480,7 @@ This option should be used when specifying the certificate via -CertificateSubje
     <value>Comma-separated list of valid authentication types for this source. By default, all authentication types are valid. Example: basic,negotiate</value>
     <comment>Please don't localize "basic,negotiate"</comment>
   </data>
+  <data name="PackageCommandSymbolPackageFormat" xml:space="preserve">
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:5483]: nit: When...

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs [hunk 18]
```diff
@@ -78,17 +78,20 @@ public PackageBuilder CreateBuilder(string basePath, NuGetVersion version, strin
...
             AddContentFiles(builder);
-            
+
             // Add sources if this is a symbol package
             if (IncludeSymbols)
             {
                 AddSourceFiles();
             }
+            
+
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: nit: delete these extra spaces.

File: src/NuGet.Core/NuGet.Commands/Strings.resx [hunk 19]
```diff
@@ -700,4 +700,7 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
     <value>Clearing NuGet plugins cache: {0}</value>
     <comment>{0}: Plugins cache path</comment>
   </data>
+  <data name="Error_InvalidSymbolPackageFormat" xml:space="preserve">
+    <value>The provided SymbolPackageFormat value {0} is invalid. Possible allowed values are either 'snupkg' or 'symbols.nupkg'.</value>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/Strings.resx]: Alllowed values: ....

similar to what we do in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx#L207-L210

Also add a comment telling not to localize snpukg and symbols.nupg. 

I would use the constant as a parameter to this error message to minimize potential future mistakes.

File: src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs [hunk 20]
```diff
@@ -90,5 +90,29 @@ public static async Task<PackageUpdateResource> GetPackageUpdateResource(IPackag
 
             return await sourceRepository.GetResourceAsync<PackageUpdateResource>();
         }
+
+        public static async Task<SymbolPackageUpdateResourceV3> GetSymbolPackageUpdateResource(IPackageSourceProvider sourceProvider, string source)
+        {
+            // Use a loaded PackageSource if possible since it contains credential info
+            PackageSource packageSource = null;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs]: You can simplify this with a linq clause. 

> sourceProvider.LoadPackageSources().Where(e => e.IsEnabled && source == loadedPackageSource.source).FirstOrDefault();

Should you use case insensitive comparison when looking for the source?

File: src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+            if (serviceIndex != null)
+            {
+                var baseUrl = serviceIndex.GetServiceEntryUri(ServiceTypes.SymbolPackagePublish);
+
+                HttpSource httpSource = null;
+                var sourceUri = baseUrl?.AbsoluteUri;
+                if (!string.IsNullOrEmpty(sourceUri))
+                {
+                    if (!(new Uri(sourceUri)).IsFile)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs:36]: Why is this necesarry? 

You're not going to get here becuase the serviceindexresource will never be returned for a file.

**Review by nkolev92 — APPROVED** [review]
Some small comments/suggestions. 

Looks great otherwise.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 3 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:986]: I know this is by design but it feels weird to silently ignore these

**Comment by jainaashish** [comment]
`like i said above, if there is a symbol.nupkg, the only way it will be pushed is if a symbol source and apikey is specified explicitly. If that is specified explicitly, we will look for symbol.nupkg instead of snupkg. For pushing a snupkg, you do not need to specify symbol source or an api key.`

I was under the impression that we're still pushing symbol packages implicitly to SymbolServer because that's what even we use in our publish workflow. But didn't realize that we're stuck with 4.3  may be for the same reason.

So let me get this straight, until 4.5 we were pushing symbol packages implicitly to SymbolServer then from nuget.exe 4.5 we changed that to only push when explicit symbol server and api key is mentioned. And now from nuget.exe 4.9 we'll start pushing snupkg implicitly to NuGet symbol server.

**Review by jainaashish — APPROVED** [review]

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs (hunk 11 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs:53]: it's the same as the main source's apiKey, i am just copying it here so i don't end up changing too many function definitinons,.

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs (hunk 14 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs:159]: we allow both.  
to push symbol.nupkg , you need to explicitly specify symbolSoruce in which case symbolPackageUpdateResource will always be null since we will not initialize it.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 16 above)
Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:822]: i think for now this is fine. if we do end up having a third option, we will modify it then.

File: src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs (hunk 21 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs:36]: this is just inspired from https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Providers/PackageUpdateResourceV3Provider.cs#L38

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs:36]: I'd remove it both places because of the following :) 

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Providers/ServiceIndexResourceV3Provider.cs#L49

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/SymbolPackageUpdateResourceV3Provider.cs:36]: synced up offline. 

This is checking whether the symbols package publish is valid.
